### PR TITLE
Feature/refactor substate serialization

### PR DESCRIPTION
**Pull request by talekhinezh**
* Refactor substate serialization/deserialization so that it is no longer maintained in a static file but is instead defined by ConstraintScrypts

### CODE REVIEW

**Review by siy — APPROVED** [review]
LGTM

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/api/construction/LogCMStore.java [hunk 1]
```diff
@@ -61,40 +66,39 @@ public boolean isVirtualDown(Transaction dbTxn, SubstateId substateId) {
...
-							if (i.getMicroOp() != REInstruction.REMicroOp.UP) {
-								return Optional.empty();
-							}
-							Substate s = i.getData();
-							return Optional.of(s.getParticle());
-						}
-					} catch (DeserializeException e) {
-						throw new IllegalStateException("Cannot deserialize instruction@" + cur, e);
+				try {
+					var instructions = reParser.parse(txn).instructions();
```
Comment (iamyulong) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/api/construction/LogCMStore.java:80]: As an optimisation opportunity for future, we might consider building index for up particles so no full transaction parsing is required.

File: radixdlt-engine/src/main/java/com/radixdlt/atommodel/tokens/scrypt/TokensConstraintScryptV1.java [hunk 2]
```diff
@@ -53,18 +58,89 @@ public void main(Loader os) {
...
 
-		os.particle(
-			TokensInAccount.class,
-			ParticleDefinition.<TokensInAccount>builder()
-				.staticValidation(TokenDefinitionUtils::staticCheck)
-				.build()
+		os.substate(
+			new SubstateDefinition<>(
+				TokensInAccount.class,
+				Set.of(SubstateTypeId.TOKENS.id(), SubstateTypeId.TOKENS_LOCKED.id()),
```
Comment (iamyulong) [inline radixdlt-engine/src/main/java/com/radixdlt/atommodel/tokens/scrypt/TokensConstraintScryptV1.java:117]: This is very personal - I didn't like the idea of reusing the same class for different types of substate, even though they share similar data structure; it introduces complexity with little benefits.

File: radixdlt-engine/src/main/java/com/radixdlt/constraintmachine/REInstruction.java [hunk 3]
```diff
@@ -31,95 +31,119 @@
...
+				var classId = b.get();
+				return d.byteToClass(classId); // Just to check to make sure classId exists
+			}
+		},
+		MSG((byte) 6, REOp.MSG) {
+			@Override
+			public Object read(Txn txn, int i, ByteBuffer b, SubstateDeserialization d) throws DeserializeException {
+				var length = Byte.toUnsignedInt(b.get());
+				var bytes = new byte[length];
+				b.get(bytes);
```
Comment (iamyulong) [inline radixdlt-engine/src/main/java/com/radixdlt/constraintmachine/REInstruction.java:93]: We may want to create a reusable method for decoding variable length byte array. It seems to be commonly used.

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/store/berkeley/BerkeleyLedgerEntryStore.java [hunk 4]
```diff
@@ -319,7 +320,7 @@ private SecondaryConfig buildUpParticleConfig() {
 		return (SecondaryConfig) new SecondaryConfig()
 			.setKeyCreator(
 				(secondary, key, data, result) -> {
-					if (entryToSpin(data) == REOp.DOWN) {
+					if (entryToSpin(data) != REOp.UP) {
```
Comment (iamyulong) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/store/berkeley/BerkeleyLedgerEntryStore.java:323]: Nice catch!

**Review by iamyulong — COMMENTED** [review]
LGTM

**Review by michalzajda — APPROVED** [review]

File: radixdlt-engine/src/main/java/com/radixdlt/atommodel/tokens/scrypt/TokensConstraintScryptV1.java (hunk 2 above)
Comment (talekhinezh) [reply radixdlt-engine/src/main/java/com/radixdlt/atommodel/tokens/scrypt/TokensConstraintScryptV1.java:117]: Yes, totally agree. In this case though, TOKENS_LOCKED is going to be deprecated so would rather keep the nastiness in one place and remove it come mainnet as its not just similar data structure but also similar state machines where these two need to be handled the same and we cant use interfaces for substates at the moment.

**Pull request merged**